open_prime_rando/echoes/dock_lock_rando/dock_type.py

```diff
@@ -1,16 +1,323 @@
+        return self.patched_scan
+
+    def patch_door(self, editor: PatcherEditor, world_name: str, area_name: str, dock_name: str):
+        mrea, mapa = self.get_files(editor, world_name, area_name)
+        door = self.get_door_from_dock_index(mrea, self.get_dock_index(world_name, area_name, dock_name))
+        self.patch_map_icon(mapa, door)
+
+        shell_model = editor._resolve_asset_id(self.shell_model)
+
+        door_props = door.get_properties_as(Door)
+        door_props.shell_model = shell_model
+        door_props.shell_color = self.shell_color
+        door.set_properties(door_props)
+
+        for pak in self.get_paks(editor, world_name, area_name):
+            editor.ensure_present(pak, shell_model)
+
+
+@dataclasses.dataclass(kw_only=True)
+class NormalDoorType(DoorType):
+    def patch_door(self, editor: PatcherEditor, world_name: str, area_name: str, dock_name: str):
+        super().patch_door(editor, world_name, area_name, dock_name)
+        mrea = self.get_mrea(editor, world_name, area_name)
+        door = self.get_door_from_dock_index(mrea, self.get_dock_index(world_name, area_name, dock_name))
+        
+        door_props = door.get_properties_as(Door)
+        door_props.vulnerability = self.vulnerability
+
+        if self.scan_text is not None:
+            door_props.alt_scannable.scannable_info0 = self.get_patched_scan(editor, world_name, area_name)
+        
+        door.set_properties(door_props)
+    
+
+@dataclasses.dataclass(kw_only=True)
+class BlastShieldDoorType(DoorType):
+    shield_model: NameOrAssetId
+    shield_collision_box: Vector = dataclasses.field(default_factory=lambda: Vector(0.35, 5.0, 4.0))
+    shield_collision_offset: Vector = dataclasses.field(default_factory=lambda: Vector(-2/3, 0, 2.0))
+    
+    def get_spline(editor: PatcherEditor) -> Spline:
+        return Spline(data=b'\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x00\x02\x02A \x00\x00?\x80\x00\x00\x02\x02\x01\x00\x00\x00\x00?\x80\x00\x00')
+
+    def patch_door(self, editor: PatcherEditor, world_name: str, area_name: str, dock_name: str):
+        """
+        blast shield connections:
+            DEAD -> lock cleared MemoryRelay, ACTV
+            DEAD -> lock breaking SoundEffect, PLAY
+            DEAD -> lock explosion gibs, ACTV
+            DEAD -> jingle streamedaudio, PLAY
+        
+        door connections:
+            OPEN -> lock cleared MemoryRelay, ACTV
+        
+        memory relay connections:
+            ACTV -> door, RSET
+            ACTV -> blast shield, DCTV
+        """
+        super().patch_door(editor, world_name, area_name, dock_name)
+        mrea = self.get_mrea(editor, world_name, area_name)
+        _door = self.get_door_from_dock_index(mrea, self.get_dock_index(world_name, area_name, dock_name))
+        door = _door.get_properties_as(Door)
+
+        default = mrea.get_layer("Default")
+        
+        _sound = default.add_instance("SOND", "Metal Door Lock Breaking")
```

Comment:
Wouldn't it be better to use `add_instance_with`? Could be something like this:
```python
sound = default.add_instance_with(Sound(
    editor_properties=EditorProperties(
        name="Metal Door Lock Breaking",
        transform=door.editor_properties.transform,
    ),
    sound=948,
    max_audible_distance=150.0,
    surround_pan=SurroundPan(surround_pan=1.0),
))
```

Reply:
makes sense. I'll need to leave the memory relay as-is tho